### PR TITLE
Update dockerfile to fix issues with mpi4py install

### PR DESCRIPTION
**Pull request by ktangsali**
# Modulus Pull Request

## Description
This PR attempts to fix the issues with mpi4py build during the docker build. It also fixes version for pip and setuptools and also adds `--no-cache-dir` flag to all the pip commands. 

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -18,7 +18,7 @@ FROM ${BASE_CONTAINER} as builder
 ARG TARGETPLATFORM
 
 # Update pip and setuptools
-RUN pip install --upgrade pip setuptools  
+RUN pip install "pip==23.2.1" "setuptools==68.2.2"
```
Comment (NickGeneva) [inline Dockerfile:21]: Curious why fixing these versions?

Comment (ktangsali) [reply Dockerfile:21]: I think its good to fix them eitherways, plus it seems like these versions are required for mpi4py

File: Dockerfile [hunk 2]
```diff
@@ -141,10 +145,10 @@ ARG TARGETPLATFORM
 RUN pip install "protobuf==3.20.3"
```
Comment (NickGeneva) [inline Dockerfile:145]: --no-cache-dir

**Review by NickGeneva — APPROVED** [review]

File: Dockerfile (hunk 2 above)
Comment (ktangsali) [reply Dockerfile]: Done

**Review by mnabian — APPROVED** [review]

**Pull request merged**